% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/coverageFromBedGraph.R
\name{coverageFromBedGraph}
\alias{coverageFromBedGraph}
\title{Read coverage from bedGraph files}
\usage{
coverageFromBedGraph(
  bedgraphs,
  tags,
  genome,
  removeScaffolds = FALSE,
  hugeData = FALSE,
  BPPARAM = NULL,
  ...
)
}
\arguments{
\item{bedgraphs}{The file names of bedgraphs generated by bedtools. 
eg: bedtools genomecov-bg -split -ibam $bam > $bedgraph}

\item{tags}{Name tags used to label each input bedgraph}

\item{genome}{An object of [BSgenome::BSgenome-class]}

\item{removeScaffolds}{logical(1), whether the scaffolds should be 
removed from the genome}

\item{hugeData}{logical(1). Is this dataset consume too much memory? 
if it is TRUE, the coverage will be saved into tempfiles.}

\item{BPPARAM}{an optional [BiocParallel::BiocParallelParam-class] 
instance determining the parallel back-end to be used during evaluation, 
or a list of BiocParallelParam instances, to be applied in sequence for
nested calls to bplapply}

\item{...}{Parameters can be passed into the tempfile() function. This is 
useful when you submit hugedataset to cluster.}
}
\value{
A list of read coverage for each bedgraph file. For each item 
 in the list, it is a list of coverage for each chromosome. And the 
 chromosome must start with "chr".
}
\description{
Read in read coverage from bedGraph files and store it as a list
}
\examples{
if(interactive()) {
library(BSgenome.Mmusculus.UCSC.mm10)
path <- file.path(find.package("InPAS"), "extdata")
bedgraphs <- file.path(path, "Baf3.extract.bedgraph")
data(utr3.mm10)
tags <- "Baf3"
genome <- BSgenome.Mmusculus.UCSC.mm10
coverage <- coverageFromBedGraph(bedgraphs, tags, 
genome, hugeData = FALSE)
}


}
