% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/inPASWrapper.R
\name{inPASWrapper}
\alias{inPASWrapper}
\title{do estimation of alternative polyadenylation and cleavage 
site in one step}
\usage{
inPASWrapper(
  bedgraphs,
  genome,
  removeScaffolds = FALSE,
  utr3,
  txdb = NA,
  tags,
  hugeData = FALSE,
  ...,
  gp1,
  gp2,
  window_size = 100,
  search_point_START = 50,
  search_point_END = NA,
  cutStart = window_size,
  cutEnd = 0,
  coverage_threshold = 5,
  long_coverage_threshold = 2,
  background = c("same_as_long_coverage_threshold", "1K", "5K", "10K", "50K"),
  adjust_distal_polyA_end = TRUE,
  PolyA_PWM = NA,
  classifier = NA,
  classifier_cutoff = 0.8,
  shift_range = window_size,
  method = c("limma", "fisher.exact", "singleSample", "singleGroup"),
  normalize = c("none", "quantiles", "quantiles.robust", "mean", "median"),
  design,
  contrast.matrix,
  coef = 1,
  P.Value_cutoff = 0.05,
  adj.P.Val_cutoff = 0.05,
  dPDUI_cutoff = 0.3,
  PDUI_logFC_cutoff = 0.59,
  BPPARAM = NULL
)
}
\arguments{
\item{bedgraphs}{The file names of bedgraphs generated by 
bedtools. eg: bedtools genomecov -bg -split -ibam $bam 
-g mm10.size.txt > $bedgraph}

\item{genome}{an object of [BSgenome::BSgenome-class]}

\item{removeScaffolds}{logical(1), whether the scaffolds should be 
removed from the genome}

\item{utr3}{output of utr3Annotation}

\item{txdb}{an object of [GenomicFeatures::TxDb-class]}

\item{tags}{the names for each input bedgraphs}

\item{hugeData}{is s this dataset consume too much memory? 
if it is TRUE, the coverage will be saved into tempfiles.}

\item{...}{parameters can be passed into tempfile. This is 
useful when you submit huge dataset to cluster.}

\item{gp1}{tag names involved in group 1}

\item{gp2}{tag names involved in group 2}

\item{window_size}{window size for novel distal position
searching and adjusted polyA searching, default: 100}

\item{search_point_START}{start point for searching}

\item{search_point_END}{end point for searching}

\item{cutStart}{how many nucleotides should be removed from 
the start before search, 0.1 means 10 percent.}

\item{cutEnd}{how many nucleotides should be removed 
from the end before search, 0.1 means 10 percent.}

\item{coverage_threshold}{cutoff threshold for coverage in the 
region of short form}

\item{long_coverage_threshold}{cutoff threshold for coverage 
in the region of long form}

\item{background}{the range for calculating cutoff threshold 
of local background}

\item{adjust_distal_polyA_end}{If true, adjust distal polyA 
end by [cleanUpdTSeq::cleanUpdTSeq-package]}

\item{PolyA_PWM}{Position Weight Matrix of polyA}

\item{classifier}{An object of [cleanUpdTSeq::PASclassifier-class]}

\item{classifier_cutoff}{This is the cutoff used to assign 
whether a putative pA is true or false. This can be any floating 
point number between 0 and 1. For example, classifier_cutoff = 
0.5 will assign an putative pA site with prob.1 > 0.5 to 
the True class (1), and any putative pA site with prob.1 <= 
0.5 as False (0).}

\item{shift_range}{the shift range for polyA site searching}

\item{method}{a character(1) vector. The available test method are #' "limma", "fisher.exact", "singleSample", "singleGroup"}

\item{normalize}{normalization method. It can be "none", "quantiles",
"quantiles.robust","mean", or "median".}

\item{design}{the design matrix of the experiment, with rows 
corresponding to arrays and columns to coefficients to be 
estimated. Defaults to the unit vector meaning that the arrays 
are treated as replicates. See [stats::model.matrix()]}

\item{contrast.matrix}{numeric matrix with rows corresponding to 
coefficients in fit and columns containing contrasts. May be a 
vector if there is only one contrast. See [limma::makeContrasts()]}

\item{coef}{column number or column name specifying which 
coefficient or contrast of the linear model is of interest. 
see more [limma::topTable()]. Default value: 1}

\item{P.Value_cutoff}{cutoff of P value}

\item{adj.P.Val_cutoff}{cutoff value for adjusted p.value}

\item{dPDUI_cutoff}{cutoff value for differential 
PAS (polyadenylation signal) usage index}

\item{PDUI_logFC_cutoff}{cutoff value for log2 fold change of 
PAS(polyadenylation signal) usage index}

\item{BPPARAM}{An optional 
[BiocParallel::BiocParallelParam-class] 
instance determining the parallel back-end to be used during 
evaluation, or a list of BiocParallelParam instances, to be 
applied in sequence for nested calls to bplapply}
}
\value{
an object of [GenomicRanges::GRanges-class]
}
\description{
A wrapper function to discover alternative polyadenylation and cleavage 
site in one step
}
\examples{
if(interactive()){
library(BSgenome.Mmusculus.UCSC.mm10)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
path <- file.path(find.package("InPAS"), "extdata")
bedgraphs <- file.path(path, "Baf3.extract.bedgraph")
data(utr3.mm10)
res <- inPASWrapper(bedgraphs=bedgraphs, tags=c("Baf3"), 
             genome=BSgenome.Mmusculus.UCSC.mm10,
             method= "limma",
             utr3=utr3.mm10, gp1="Baf3", gp2=NULL,
             txdb=TxDb.Mmusculus.UCSC.mm10.knownGene,
             search_point_START=200,
             short_coverage_threshold=15,
             long_coverage_threshold=3, 
             cutStart=0, cutEnd=.2, 
             hugeData=FALSE)
res
}
}
